### PR TITLE
Do not require UIA when first uploading cross-signing keys

### PR DESCRIPTION
**Pull request by zecakeh**
As per MSC3967.

Preview: https://pr1828--matrix-spec-previews.netlify.app

### CODE REVIEW

File: data/api/client-server/cross_signing.yaml [hunk 1]
```diff
@@ -19,11 +19,26 @@ paths:
...
 
         This API endpoint uses the [User-Interactive Authentication API](/client-server-api/#user-interactive-authentication-api).
+        
+        User-Interactive Authentication is required, except in these cases:
+        - there is no existing cross-signing master key uploaded to the homeserver, OR
+        - there is an existing cross-signing master key and it exactly matches the
+          cross-signing master key provided in the request body. If there are any additional
+          keys provided in the request (self-signing key, user-signing key) they MUST also
+          match the existing keys stored on the server. In other words, the request contains
+          no new keys.
```
Comment (zecakeh) [inline data/api/client-server/cross_signing.yaml:36]: I copied this section from the MSC, but it seems to ignore the case where a client uploads the master key and the other keys (self-signing and user-signing) separately.

My understanding according to the paragraph below is that setting the other keys if they were not set before, or re-uploading the same keys, should not require UIA, even if the master key is not part of the request (i.e. it was already uploaded before).

**Comment by richvdh** [comment]
LGTM other than a couple of nits.

File: data/api/client-server/cross_signing.yaml [hunk 2]
```diff
@@ -19,11 +19,26 @@ paths:
...
+        
+        This allows clients to freely upload 1 set of keys, but not modify/overwrite keys if
```
[the remark this answers is not in the document]
Comment (zecakeh) [reply data/api/client-server/cross_signing.yaml]: Applied

File: data/api/client-server/cross_signing.yaml [hunk 3]
```diff
@@ -19,11 +19,26 @@ paths:
...
+        they already exist. By allowing clients to upload the same set of keys more than once,
+        this makes this endpoint idempotent in the case where the response is lost over the
```
[the remark this answers is not in the document]
Comment (zecakeh) [reply data/api/client-server/cross_signing.yaml]: Applied

File: data/api/client-server/cross_signing.yaml [hunk 4]
```diff
@@ -19,11 +19,26 @@ paths:
...
       x-addedInMatrixVersion: "1.1"
+      x-changedInMatrixVersion:
+        "1.11": UIA is not always required for this endpoint.
       summary: Upload cross-signing keys.
       description: |-
         Publishes cross-signing keys for the user.
 
         This API endpoint uses the [User-Interactive Authentication API](/client-server-api/#user-interactive-authentication-api).
+        
+        User-Interactive Authentication is required, except in these cases:
```
[the remark this answers is not in the document]
Comment (zecakeh) [reply data/api/client-server/cross_signing.yaml]: Applied

**Review by richvdh — APPROVED** [review]
Thanks!

**Pull request merged**